Add AGS10 Sensor

mak-42 opened the pull request:
Add support for AGS10 VOC Sensor.

<!-- Quick description and explanation of changes -->

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3527

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
    i2c:
      sda: GPIO21
      scl: GPIO22
      scan: true
      frequency: 15kHz

    sensor:
      - platform: ags10
        tvoc:
          name: "TVOC"
        address: 0x1A
        update_interval: 60s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

mak-42 committed:
Add AGS10 Sensor

mak-42 committed:
Fix issues with  AGS10 Sensor implementation code.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.
Add tests.

mak-42 committed:
Fix issues with AGS10 Sensor test code.

mak-42 committed:
Fix issues with AGS10 Sensor test code.

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,122 @@
+            ),
+            cv.Optional(CONF_RESISTANCE): sensor.sensor_schema(
+                icon=ICON_RESTART,
+            ),
```
This sensor should be setting most if not all of the arguments to `sensor_schema` and also have `entity_category=ENTITY_CATEGORY_DIAGNOSTIC` (The `version` sensor should also be a diagnostic sensor)

mak-42 replied:
Done. Please check if it is enough.  The devices which code I used as example does not contain that :(

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(AGS10Component),
+            cv.Required(CONF_TVOC): sensor.sensor_schema(
```
All sensors are to be Optional
```suggestion
            cv.Optional(CONF_TVOC): sensor.sensor_schema(
```

mak-42 replied:
Done

jesserockz commented on esphome/components/ags10/ags10.h:
```diff
@@ -0,0 +1,160 @@
+  CUSTOM_VALUE,
+};
+
+template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...> {
```
After adding this, you can remove the constructor and protected field
```suggestion
template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...>, public Parented<AGS10Component> {
```

mak-42 replied:
Done

mak-42 replied:
I've to leave constructors, because the compilation error occurs without it:
```
src/main.cpp: In function 'void setup()':
src/main.cpp:1274:81: error: no matching function for call to 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10newi2caddressaction = new ags10::AGS10NewI2cAddressAction<>(ags10_1);
                                                                                 ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction()'
 template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:116:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(const esphome::ags10::AGS10NewI2cAddressAction<>&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10NewI2cAddressAction<>&'
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10NewI2cAddressAction<>&&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10NewI2cAddressAction<>&&'
src/main.cpp:1301:83: error: no matching function for call to 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10newi2caddressaction_2 = new ags10::AGS10NewI2cAddressAction<>(ags10_1);
                                                                                   ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction()'
 template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:116:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(const esphome::ags10::AGS10NewI2cAddressAction<>&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10NewI2cAddressAction<>&'
src/esphome/components/ags10/ags10.h:116:32: note: candidate: 'esphome::ags10::AGS10NewI2cAddressAction<>::AGS10NewI2cAddressAction(esphome::ags10::AGS10NewI2cAddressAction<>&&)'
src/esphome/components/ags10/ags10.h:116:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10NewI2cAddressAction<>&&'
src/main.cpp:1329:79: error: no matching function for call to 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10setzeropointaction = new ags10::AGS10SetZeroPointAction<>(ags10_1);
                                                                               ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction()'
 template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:132:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(const esphome::ags10::AGS10SetZeroPointAction<>&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10SetZeroPointAction<>&'
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10SetZeroPointAction<>&&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10SetZeroPointAction<>&&'
src/main.cpp:1357:81: error: no matching function for call to 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10Component*&)'
   ags10_ags10setzeropointaction_2 = new ags10::AGS10SetZeroPointAction<>(ags10_1);
                                                                                 ^
In file included from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction()'
 template<typename... Ts> class AGS10SetZeroPointAction : public Action<Ts...>, public Parented<AGS10Component> {
                                ^~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ags10/ags10.h:132:32: note:   candidate expects 0 arguments, 1 provided
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(const esphome::ags10::AGS10SetZeroPointAction<>&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'const esphome::ags10::AGS10SetZeroPointAction<>&'
src/esphome/components/ags10/ags10.h:132:32: note: candidate: 'esphome::ags10::AGS10SetZeroPointAction<>::AGS10SetZeroPointAction(esphome::ags10::AGS10SetZeroPointAction<>&&)'
src/esphome/components/ags10/ags10.h:132:32: note:   no known conversion for argument 1 from 'esphome::ags10::AGS10Component*' to 'esphome::ags10::AGS10SetZeroPointAction<>&&'
```

mak-42 replied:
Finally I managed to remove constructors.

jesserockz commented on esphome/components/ags10/ags10.h:
```diff
@@ -0,0 +1,160 @@
+template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...> {
+ public:
+  AGS10NewI2cAddressAction(AGS10Component *parent) : parent_(parent) {}
+  TEMPLATABLE_VALUE(int, new_address)
```
I2C addresses are limites to a single byte, no need to reserve more memory that wont be used
```suggestion
  TEMPLATABLE_VALUE(uint8_t, new_address)
```

mak-42 replied:
Done

jesserockz commented on esphome/components/ags10/ags10.h:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * Sets TVOC sensor.
+   */
+  void set_tvoc(sensor::Sensor *tvoc) { tvoc_ = tvoc; }
```
Please make sure to always prefix member variables and function calls with `this->` as per our [code standards](https://esphome.io/guides/contributing#codebase-standards)
```suggestion
  void set_tvoc(sensor::Sensor *tvoc) { this->tvoc_ = tvoc; }
```

mak-42 replied:
Done

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,122 @@
+)
+AGS10SetZeroPointAction = ags10_ns.class_("AGS10SetZeroPointAction", automation.Action)
+
+CONFIG_SCHEMA = (
```
Please also look at `final_validate_device_schema` and look in the codebase on how to use it to specify a `max_frequency` of the i2c bus.

mak-42 replied:
Done. 
I've also move tests to separated files, because configuration i2c of all the tests has higher frequency then required.

mak-42 committed:
Merge branch 'dev' into ags10

mak-42 committed:
Fix the issues pointed by reviewer:
1. TVOC sensor now is optional.
2. "sensor_schema" of diagnostic sensors was filled properly.
3. Max I2C frequency was defined in validation schema.
4. Definition of templates were simplified.
5. "this->" was added in setting class fields.
6. Proper type was set for i2c address.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
New test for AGS10 was created because it required low I2C frequency.

mak-42 committed:
New test for AGS10 was created because it required low I2C frequency (file name correction).

mak-42 committed:
Fix issues with AGS10 Sensor implementation code.

mak-42 committed:
Merge branch 'dev' into ags10

mak-42 committed:
Merge branch 'dev' into ags10

mak-42 committed:
Merge branch 'dev' into ags10

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,131 @@
+)
+async def ags10setzeropoint_to_code(config, action_id, template_arg, args):
+    parent = await cg.get_variable(config[CONF_ID])
+    var = cg.new_Pvariable(action_id, template_arg, parent)
```
```suggestion
    var = cg.new_Pvariable(action_id, template_arg)
    await cg.register_parented(var, config[CONF_ID])
```
```

mak-42 replied:
Done. It allowed to remove constructors as you suggested previously.

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,131 @@
+    var = cg.new_Pvariable(action_id, template_arg, parent)
+    var.set_mode(await cg.templatable(config.get(CONF_MODE), args, enumerate))
+    if value := config.get(CONF_VALUE):
+        cg.add(var.set_value(await cg.templatable(value, args, int)))
```
Since there is a default (`0xFFFF`), the value is always in config
```suggestion
    value = await cg.templatable(config[CONF_VALUE], args, int)
    cg.add(var.set_value(value))
```

mak-42 replied:
Done.

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,131 @@
+    "CUSTOM_VALUE": AGS10SetZeroPointActionMode.CUSTOM_VALUE,
+}
+
+AGS10_SET_ZERO_POINT_SCHEMA = cv.All(
```
```suggestion
AGS10_SET_ZERO_POINT_SCHEMA = cv.Schema(
```

mak-42 replied:
Done

jesserockz commented on esphome/components/ags10/sensor.py:
```diff
@@ -0,0 +1,131 @@
+async def ags10setzeropoint_to_code(config, action_id, template_arg, args):
+    parent = await cg.get_variable(config[CONF_ID])
+    var = cg.new_Pvariable(action_id, template_arg, parent)
+    var.set_mode(await cg.templatable(config.get(CONF_MODE), args, enumerate))
```
Mode is not templatable in the above SCHEMA...

mak-42 replied:
Fixed.

mak-42 committed:
Merge branch 'dev' into ags10

mak-42 committed:
Fix the issues pointed by reviewer:
1. "cg.register_parented" is using now.
2. Schema was corrected.
3. "Mode" now is templatable.

mak-42 committed:
Fix the issues pointed by reviewer:
1. Constructors were removed.
